### PR TITLE
update onehot, cifar apphub

### PR DESCRIPTION
**Pull request by vbvg2008**

### CODE REVIEW

File: fastestimator/op/numpyop/univariate/onehot.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class Onehot(NumpyOp):
+    """ Transform the label integer to one-hot-encoding
+        Ref: https://towardsdatascience.com/label-smoothing-making-model-robust-to-incorrect-labels-2fae037ffbd0
+
+    Args:
+        num_classes: Total number of classes.
+        label_smoothing: Smoothing factor. The class index value will become: 1 - label_smoothing + label_smoothing
+            / num_classes and The other class index value will become label_smoothing / num_classes
```
[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/op/numpyop/univariate/onehot.py]: nice

File: fastestimator/backend/get_gradient.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        gradients as sequence of tensors
+    """
+    if tape:
+        with NonContext() if higher_order else tape.stop_recording():
+            gradients = tape.gradient(target, sources)
+    else:
+        gradients = torch.autograd.grad(target,
+                                        sources,
+                                        retain_graph=retain_graph,
+                                        create_graph=higher_order,
```
Comment (TortoiseHam) [inline fastestimator/backend/get_gradient.py]: not sure that create_graph = True will result in higher order derivatives. When I run two calls with that flag set the result is the same each time

File: fastestimator/backend/get_gradient.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# ==============================================================================
+from typing import Iterable, Optional, Union
+
+import tensorflow as tf
+import torch
+
+from fastestimator.util.util import NonContext
+
+
+def get_gradient(target: Union[tf.Tensor, torch.Tensor],
```
Comment (TortoiseHam) [inline fastestimator/backend/get_gradient.py]: Lets use the Tensor TypeVar here

File: fastestimator/backend/get_gradient.py (hunk 2 above)
Comment (vbvg2008) [reply fastestimator/backend/get_gradient.py]: it should be the same results, but when create_graph=true, the gradient output will have a .grad field - that's what we are looking for when doing higher order

File: fastestimator/op/numpyop/univariate/onehot.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        num_classes: Total number of classes.
+        label_smoothing: Smoothing factor, after smoothing class output is: 1 - label_smoothing + label_smoothing
+            / num_classes, the other class output is: label_smoothing / num_classes
+        inputs: Input key(s) of labels to be onehot encoded
+        outputs: Output key(s) of labels
+        mode: What execution mode (train, eval, None) to apply this operation
+    """
+    def __init__(self,
+                 num_classes: int,
+                 label_smoothing: Optional[float] = None,
```
Comment (TortoiseHam) [inline fastestimator/op/numpyop/univariate/onehot.py]: instead of making this an optional float why not just default to 0?

File: fastestimator/op/numpyop/univariate/onehot.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                 outputs: Union[None, str, Iterable[str]] = None,
+                 mode: Union[None, str, Iterable[str]] = None):
+
+        super().__init__(inputs=inputs, outputs=outputs, mode=mode)
+        self.num_classes = num_classes
+        self.label_smoothing = label_smoothing
+        if not self.label_smoothing:
+            self.label_smoothing = 0.0
+
+    def forward(self, data: Union[int, np.ndarray], state: Dict[str, Any]) -> Union[np.ndarray, List[np.ndarray]]:
```
Comment (TortoiseHam) [inline fastestimator/op/numpyop/univariate/onehot.py]: this should be written to support multi-io since it's univariate and 'inputs' method signature indicates it accepts Iterable[str] as inputs

File: fastestimator/op/numpyop/univariate/onehot.py (hunk 4 above)
Comment (vbvg2008) [reply fastestimator/op/numpyop/univariate/onehot.py]: good idea, done

File: fastestimator/backend/get_gradient.py (hunk 3 above)
Comment (vbvg2008) [reply fastestimator/backend/get_gradient.py]: done

File: fastestimator/backend/get_gradient.py (hunk 2 above)
Comment (TortoiseHam) [reply fastestimator/backend/get_gradient.py]: when i run an example of this the result tensor .grad is not populated either

File: fastestimator/op/numpyop/univariate/onehot.py (hunk 5 above)
Comment (vbvg2008) [reply fastestimator/op/numpyop/univariate/onehot.py]: multi-IO needs further debate, let's address this later.
